src/main/java/gyro/google/compute/TargetPoolResource.java

```diff
@@ -0,0 +1,291 @@
+    @Output
+    public String getSelfLink() {
+        return selfLink;
+    }
+
+    public void setSelfLink(String selfLink) {
+        this.selfLink = selfLink;
+    }
+
+    @Override
+    public void copyFrom(TargetPool model) throws Exception {
+        setBackupPool(findById(TargetPoolResource.class, model.getBackupPool()));
+        setDescription(model.getDescription());
+        setFailoverRatio(model.getFailoverRatio());
+        setName(model.getName());
+        setSessionAffinity(model.getSessionAffinity());
+        setRegion(model.getRegion());
+        setSelfLink(model.getSelfLink());
+
+        List<String> instances = model.getInstances();
+        if (instances != null) {
+            setInstances(instances.stream().map(e -> findById(InstanceResource.class, e)).collect(Collectors.toList()));
+        }
+    }
+
+    @Override
+    protected boolean doRefresh() throws Exception {
+        Compute client = createComputeClient();
+
+        TargetPool targetPool = client.targetPools().get(getProjectId(), getRegion(), getName()).execute();
+        copyFrom(targetPool);
+
+        return true;
+    }
+
+    @Override
+    protected void doCreate(GyroUI ui, State state) throws Exception {
+        TargetPool targetPool = new TargetPool();
+        targetPool.setBackupPool(getBackupPool() != null ? getBackupPool().getSelfLink() : null);
+        targetPool.setDescription(getDescription());
+        targetPool.setFailoverRatio(getFailoverRatio());
+        targetPool.setName(getName());
+        targetPool.setSessionAffinity(getSessionAffinity());
+        targetPool.setInstances(
+            getInstances().stream().map(InstanceResource::getSelfLink).collect(Collectors.toList()));
+
+        Compute client = createComputeClient();
+        Operation operation = client.targetPools().insert(getProjectId(), getRegion(), targetPool).execute();
+        waitForCompletion(client, operation);
+
+        refresh();
+    }
+
+    @Override
+    public void doUpdate(GyroUI ui, State state, Resource current, Set<String> changedFieldNames) throws Exception {
+        TargetPoolResource currentTargetPool = (TargetPoolResource) current;
+        Compute client = createComputeClient();
+
+        if (changedFieldNames.contains("backup-pool") || changedFieldNames.contains("failover-ratio")) {
+            TargetReference targetReference = new TargetReference();
+            targetReference.setTarget(getBackupPool() != null ? getBackupPool().getSelfLink() : null);
+            Operation response = client.targetPools()
+                .setBackup(getProjectId(), getRegion(), getName(), targetReference)
+                .setFailoverRatio(getFailoverRatio())
+                .execute();
+            waitForCompletion(client, response);
+        }
+
+        if (changedFieldNames.contains("instances")) {
+            List<InstanceResource> removeInstances = currentTargetPool.getInstances();
+            removeInstances.removeAll(getInstances());
```

Comment:
The `removeAll()` wont work on objects unless you override the `equals` method of the class.
The objects in the collection are similar but not the same as they are instantiated separately and so are different in nature. Overriding the `equals` method gives the control back to you to compare based on defined attributes.

In the current case, the self link is a field we can look to see if two instances are common and remove it from the add/delete operation

```suggestion
            Set<String> instances = getInstances().stream()
                .map(InstanceResource::getSelfLink)
                .collect(Collectors.toSet());
            removeInstances.removeIf(o -> instances.contains(o.getSelfLink()));
```